File: connector/connector-domparser/src/main/java/com/github/thetric/iliasdownloader/service/webparser/WebParserIliasServiceProvider.java
package com.github.thetric.iliasdownloader.service.webparser;

import com.github.thetric.iliasdownloader.service.IliasService;
import com.github.thetric.iliasdownloader.service.IliasServiceProvider;
import lombok.NonNull;
import org.jsoup.Jsoup;

import java.io.IOException;
import java.util.Optional;

import static org.jsoup.Connection.Response;

/**
 * @author broj
 * @since 25.09.2016
 */
public final class WebParserIliasServiceProvider implements IliasServiceProvider {
    private static final String LOGIN_PAGE_NAME = "login.php";
    private static final String ILIAS_CLIENT_ID_COOKIE_NAME = "ilClientId";

    private final String iliasBaseUrl;
    private final String clientId;

    public WebParserIliasServiceProvider(@NonNull String loginPage) throws IOException {
        iliasBaseUrl = getBaseUrl(loginPage);
        clientId = getClientId(loginPage);
    }

    private String getBaseUrl(String loginPage) {
        loginPage = loginPage.trim();
        if (loginPage.isEmpty()) {
            throw new IllegalArgumentException("Die angegebene Loginseiten URL darf nicht leer sein");
        }
        if (!loginPage.startsWith("http://") && !loginPage.startsWith("https://")) {
            loginPage = "https://" + loginPage;
        }
        int loginPageNameIndex = loginPage.indexOf(LOGIN_PAGE_NAME);
        if (loginPageNameIndex == -1) {
            throw new IllegalArgumentException("Die angegebene URL enthält kein '" + LOGIN_PAGE_NAME + "'");
        }
        return loginPage.substring(0, loginPageNameIndex);
    }

    private String getClientId(String loginPage) throws IOException {
        String id;
        try {
            Response response = Jsoup.connect(loginPage).execute();
            id = response.cookie(ILIAS_CLIENT_ID_COOKIE_NAME);
        } catch (IOException e) {
            throw new IOException("Konnte die URL '" + loginPage + "' nicht erreichen", e);
        }
        return Optional.ofNullable(id).orElseThrow(() -> new NoCookiesAvailableException(String.format(
                "Konnte das Cookie '%s' nicht in der Response von der Seite %s finden",
                ILIAS_CLIENT_ID_COOKIE_NAME,
                loginPage))
        );
    }

    @Override
    public IliasService newInstance() {
        return new WebIliasService(new WebIoExceptionTranslatorImpl(), iliasBaseUrl, clientId);
    }
}

File: src/main/kotlin/com/github/thetric/iliasdownloader/connector/domparser/WebParserIliasServiceProvider.kt
package com.github.thetric.iliasdownloader.connector.domparser

import com.github.thetric.iliasdownloader.connector.api.IliasService
import com.github.thetric.iliasdownloader.connector.api.IliasServiceProvider
import org.jsoup.Jsoup
import java.io.IOException

private const val LOGIN_PAGE_NAME = "login.php"
private const val ILIAS_CLIENT_ID_COOKIE_NAME = "ilClientId"

/**
 * Provides an [IliasService] that retrieves information from the Ilias service by parsing the HTML pages.
 */
class WebParserIliasServiceProvider @Throws(IOException::class)
constructor(
    loginPage: String
) : IliasServiceProvider {
    private val iliasBaseUrl: String
    private val clientId: String

    init {
        iliasBaseUrl = retrieveBaseUrl(loginPage)
        clientId = retrieveClientId(loginPage)
    }

    private fun retrieveBaseUrl(loginPage: String): String {
        var trimmed = loginPage.trim { it <= ' ' }
        require(trimmed.isNotEmpty()) {
            "Die angegebene Loginseiten URL darf nicht leer sein"
        }

        if (!trimmed.startsWith("http://") && !trimmed.startsWith("https://")) {
            trimmed = "https://$loginPage"
        }

        val loginPageNameIndex = trimmed.indexOf(LOGIN_PAGE_NAME)
        require(loginPageNameIndex != -1) {
            "Die angegebene URL enthält kein \'$LOGIN_PAGE_NAME\'"
        }

        return trimmed.substring(0, loginPageNameIndex)
    }

    @Throws(IOException::class)
    private fun retrieveClientId(loginPage: String): String {
        return try {
            val response = Jsoup.connect(loginPage).execute()
            response.cookie(ILIAS_CLIENT_ID_COOKIE_NAME)
                ?: throw CookieNotFoundException(
                    "Konnte das Cookie '$ILIAS_CLIENT_ID_COOKIE_NAME\' nicht in der "
                        + "Response von der Seite $loginPage finden"
                )
        } catch (e: IOException) {
            throw IOException("Konnte die URL \'$loginPage\' nicht erreichen", e)
        }
    }

    override fun newInstance(): IliasService {
        val iliasWebClient = IliasWebClient(iliasBaseUrl, clientId)
        val courseOverview = "${iliasBaseUrl}ilias.php?baseClass=ilPersonalDesktopGUI&cmd=jumpToSelectedItems"
        val courseLinkPrefix = "${iliasBaseUrl}goto_${clientId}_crs_"
        val courseWebDavPrefix = "${iliasBaseUrl}webdav.php/$clientId/ref_"
        val itemParser = IliasItemParser(courseWebDavPrefix, courseLinkPrefix)
        val courseSyncServiceProvider = CourseSyncService(iliasWebClient, itemParser, courseOverview)
        return WebIliasService(courseSyncServiceProvider, iliasWebClient)
    }
}
